packages/rest/test/acceptance/coercion/coercion.acceptance.ts

```diff
@@ -26,29 +31,67 @@ describe('Coercion', () => {
     createNumberFromHeader(@param.header.number('num') num: number) {
       return num;
     }
+
+    @get('/object-from-query')
+    getObjectFromQuery(@param.query.object('filter') filter: Object) {
+      return filter;
+    }
   }
 
   it('coerces parameter in path from string to number', async () => {
-    const spy = sinon.spy(MyController.prototype, 'createNumberFromPath');
+    spy = sinon.spy(MyController.prototype, 'createNumberFromPath');
     await client.get('/create-number-from-path/100').expect(200);
     sinon.assert.calledWithExactly(spy, 100);
   });
 
   it('coerces parameter in header from string to number', async () => {
-    const spy = sinon.spy(MyController.prototype, 'createNumberFromHeader');
+    spy = sinon.spy(MyController.prototype, 'createNumberFromHeader');
     await client.get('/create-number-from-header').set({num: 100});
     sinon.assert.calledWithExactly(spy, 100);
   });
 
   it('coerces parameter in query from string to number', async () => {
-    const spy = sinon.spy(MyController.prototype, 'createNumberFromQuery');
+    spy = sinon.spy(MyController.prototype, 'createNumberFromQuery');
     await client
       .get('/create-number-from-query')
       .query({num: 100})
       .expect(200);
     sinon.assert.calledWithExactly(spy, 100);
   });
 
+  it('coerces parameter in query from JSON to object', async () => {
+    spy = sinon.spy(MyController.prototype, 'getObjectFromQuery');
+    await client
+      .get('/object-from-query')
+      .query({filter: '{"where":{"id":1,"name":"Pen", "active": true}}'})
+      .expect(200);
+    sinon.assert.calledWithExactly(spy, {
+      where: {id: 1, name: 'Pen', active: true},
+    });
+  });
+
+  it('coerces parameter in query from nested keys to object', async () => {
+    spy = sinon.spy(MyController.prototype, 'getObjectFromQuery');
+    await client
+      .get('/object-from-query')
+      .query({
+        'filter[where][id]': 1,
+        'filter[where][name]': 'Pen',
+        'filter[where][active]': true,
+      })
+      .expect(200);
+    sinon.assert.calledWithExactly(spy, {
+      // Notice that numeric and boolean values are converted to strings.
```

Comment:
I don't think `qs` can handle types. It only produces strings, IIRC.

Reply:
Query strings are encoding all values as a string, there is no (standard) way how to convey type information.

Here is the request URL as sent by this test. Please ignore extra whitespace, I added it to make the example easier to read.

```
GET /object-from-query
  ?filter[where][id]=1
  &filter[where][name]=Pen
  &filter[where][active]=true
```

When this request is parsed, we get the following `request.query` object:

```js
{
  filter: {
    id: '1',
    name: 'Pen',
    active: 'true',
  }
}
```

Because we don't have any schema for the "filter" parameter, we cannot tell whether `id` should be treated as a number or as a string. There are valid cases where a value looking like a number needs to be treated as a string. MongoDB's Object IDs have taught us that lesson  back in LoopBack 2.x days.

Reply:
I see. Was wondering why the data types are preserved in the test case above but not this one, now I realized the reason...the filter in last test case is provided as a whole string, and therefore the data types are not converted to string when parsing it to a JSON object.